src/tls.c

```diff
@@ -8770,24 +8769,50 @@ int TLSX_KeyShare_Parse(WOLFSSL* ssl, const byte* input, word16 length,
         if (ssl->error != WC_PENDING_E)
     #endif
         {
-            /* Check the selected group was supported by ClientHello extensions. */
+            /* Check the selected group was supported by ClientHello
+             * extensions. */
             if (!TLSX_SupportedGroups_Find(ssl, group, ssl->extensions)) {
                 WOLFSSL_ERROR_VERBOSE(BAD_KEY_SHARE_DATA);
                 return BAD_KEY_SHARE_DATA;
             }
 
-            /* Check if the group was sent. */
-            if (TLSX_KeyShare_Find(ssl, group)) {
-                WOLFSSL_ERROR_VERBOSE(BAD_KEY_SHARE_DATA);
-                return BAD_KEY_SHARE_DATA;
+#ifdef WOLFSSL_DTLS_CH_FRAG
+            /* If we sent an empty key share then we can just limit the keyshare
+             * to the one selected by the server. */
+            if (ssl->options.dtlsSentEmptyKS) {
+                if (!TLSX_KeyShare_SelectGroup(ssl, group)) {
+                    /* Clear out all groups if not found */
+                    ret = TLSX_KeyShare_Empty(ssl);
+                    if (ret != 0)
+                        return ret;
+                }
+            }
+            else
+#endif
+            {
+                /* Check if the group was sent. */
+                if (TLSX_KeyShare_Find(ssl, group)) {
+                    WOLFSSL_ERROR_VERBOSE(BAD_KEY_SHARE_DATA);
+                    return BAD_KEY_SHARE_DATA;
+                }
+
+                /* Clear out unusable key shares. */
+                ret = TLSX_KeyShare_Empty(ssl);
+                if (ret != 0)
+                    return ret;
```

Comment:
can we generalize this improved behavior also when WOLFSSL_DTLS_CH_FRAG == 0? Also consider that RFC requires the Client to reply with just a single KeyShare after HRR.

Reply:
No because when we receive a KeyShare in HRR then that means that all of the sent keys are invalid.